File: packages/core/src/parser/__testdata__/sample.go
package synapse

import (
	"context"
	"fmt"
	"sync"
	"time"
)

const (
	MaxRetries    = 3
	DefaultModel  = "claude-opus-4-6"
	DefaultTimeout = 30 * time.Second
)

var (
	defaultConfig *Config
	mu            sync.RWMutex
)

type Status int

const (
	StatusActive Status = iota
	StatusIdle
	StatusStopped
)

type Config struct {
	Model      string `json:"model"`
	MaxTokens  int    `json:"max_tokens"`
	Temperature float64 `json:"temperature"`
}

type Agent interface {
	Process(ctx context.Context, msg string) (string, error)
	GetTools() []string
	Status() Status
}

type SynapseAgent struct {
	Name   string
	config *Config
	status Status
	mu     sync.RWMutex
}

func NewAgent(name string, cfg *Config) *SynapseAgent {
	if cfg == nil {
		cfg = &Config{Model: DefaultModel, MaxTokens: 4096}
	}
	return &SynapseAgent{Name: name, config: cfg, status: StatusIdle}
}

func (a *SynapseAgent) Process(ctx context.Context, msg string) (string, error) {
	a.mu.Lock()
	a.status = StatusActive
	a.mu.Unlock()

	defer func() {
		a.mu.Lock()
		a.status = StatusIdle
		a.mu.Unlock()
	}()

	select {
	case <-ctx.Done():
		return "", ctx.Err()
	default:
		return fmt.Sprintf("Response to: %s", msg), nil
	}
}

func (a *SynapseAgent) GetTools() []string {
	return []string{"search", "read", "write"}
}

func (a *SynapseAgent) Status() Status {
	a.mu.RLock()
	defer a.mu.RUnlock()
	return a.status
}

func init() {
	defaultConfig = &Config{Model: DefaultModel, MaxTokens: 4096, Temperature: 0.7}
}

// TODO: add graceful shutdown
// FIXME: memory leak in long-running sessions

File: packages/core/src/parser/__testdata__/sample.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc;

pub const MAX_RETRIES: u32 = 3;
pub const DEFAULT_MODEL: &str = "claude-opus-4-6";
static INSTANCE_COUNT: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(0);

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Status {
    Active,
    Idle,
    Stopped,
    Error(String),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Config {
    pub model: String,
    pub max_tokens: usize,
    pub temperature: f64,
}

pub trait Agent: Send + Sync {
    fn process(&self, message: &str) -> Result<String, AgentError>;
    fn get_tools(&self) -> Vec<String>;
    fn status(&self) -> Status;
}

pub trait Configurable {
    fn configure(&mut self, config: Config) -> Result<(), AgentError>;
}

#[derive(Debug)]
pub struct AgentError {
    pub message: String,
    pub code: u32,
}

pub struct SynapseAgent {
    name: String,
    config: Config,
    status: Arc<Mutex<Status>>,
    tools: Vec<String>,
}

impl SynapseAgent {
    pub fn new(name: &str, config: Config) -> Self {
        Self {
            name: name.to_string(),
            config,
            status: Arc::new(Mutex::new(Status::Idle)),
            tools: vec!["search".into(), "read".into()],
        }
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    fn validate_message(&self, msg: &str) -> bool {
        !msg.is_empty()
    }
}

impl Agent for SynapseAgent {
    fn process(&self, message: &str) -> Result<String, AgentError> {
        if !self.validate_message(message) {
            return Err(AgentError { message: "Empty message".into(), code: 400 });
        }
        *self.status.lock().unwrap() = Status::Active;
        let result = format!("Response to: {}", message);
        *self.status.lock().unwrap() = Status::Idle;
        Ok(result)
    }

    fn get_tools(&self) -> Vec<String> {
        self.tools.clone()
    }

    fn status(&self) -> Status {
        self.status.lock().unwrap().clone()
    }
}

impl Default for Config {
    fn default() -> Self {
        Config {
            model: DEFAULT_MODEL.to_string(),
            max_tokens: 4096,
            temperature: 0.7,
        }
    }
}

pub mod utils {
    pub fn sanitize(input: &str) -> String {
        input.trim().to_string()
    }
}

// TODO: implement async trait methods
// FIXME: status lock contention under load
